### PR TITLE
Add OTOT algo

### PR DESCRIPTION
**Pull request by trangmei**
Implements OTOT algo which applies the appropriate algo (either accelerating or decelerating) depending on the student's learning pace. Specs discussed at https://groups.google.com/d/topic/coursemology/Hu24Fjf-rlE/discussion
Fixes https://github.com/Coursemology/coursemology2/issues/3686

Add checks to prevent past personal times from being moved. Fixes https://github.com/Coursemology/coursemology2/issues/3688

### CODE REVIEW

File: app/controllers/concerns/course/lesson_plan/personalization_concern.rb [hunk 1]
```diff
@@ -28,6 +28,26 @@ def algorithm_fixed(course_user)
...
+
+    learning_rate_ema = compute_learning_rate_ema(
+      course_user, items_affects_personal_times, submitted_lesson_plan_item_ids
+    )
+    return if learning_rate_ema.nil?
+    if learning_rate_ema < 1
+      algorithm_fomo(course_user)
+    else
+      algorithm_stragglers(course_user)
+    end
```
Comment (fonglh) [inline app/controllers/concerns/course/lesson_plan/personalization_concern.rb]: `learning_rate_ema < 1 ? algorithm_fomo(course_user) : algorithm_stragglers(course_user)`

File: app/controllers/concerns/course/lesson_plan/personalization_concern.rb [hunk 2]
```diff
@@ -71,11 +91,13 @@ def algorithm_fomo(course_user)
             personal_point + (reference_time.start_at - reference_point) * learning_rate_ema,
             course_tz,
             FOMO_DATE_ROUNDING_THRESHOLD
-          )
+          ) if personal_time.start_at > Time.zone.now
```
Comment (fonglh) [inline app/controllers/concerns/course/lesson_plan/personalization_concern.rb]: Should this be based on the user's timezone or the course timezone instead?

For an example of course timezone use, see https://github.com/Coursemology/coursemology2/blob/master/app/jobs/consolidated_item_email_job.rb#L10

Comment (fonglh) [reply app/controllers/concerns/course/lesson_plan/personalization_concern.rb]: Same for the other `Time.zone.now` comparisons below.

Comment (trangmei) [reply app/controllers/concerns/course/lesson_plan/personalization_concern.rb]: This should be comparison between two datetimes with zone, so any difference in timezone should be handled by rails.

**Comment by trangmei** [comment]
Tested on local with prod data

1. Switched a leaderboard student from FOMO to OTOT, nothing changes. This is correct because the fast student's timeline is essentially still subjected to FOMO calculation.

2. Switched a student with learning rate > 1 (randomly picked a student who submitted an overdue mission) from FOMO to OTOT:

Before:
image

After:
image

This is correct because slow student's timeline is now subjected to Stragglers calculation, and we see deadline extension.

3. Switched a leaderboard student from FOMO to Straggler's algo. Assessments that were opened before remains open. Assessments that were not yet opened on FOMO get reset to reference opening date.

This is correct because personal times in the past should not change at all.
